### ISSUE TITLE
No will messages from v3.0.1 or greater

### ISSUE THREAD

**Issue by hgascoigne** [emqx/emqx#2564]
#### Environment

- Erlang/OTP: R21/10.2.1
- EMQ: > v3.0.1

#### Description

We recently attempted to upgrade from v.3.0.0 to v3.1.1 and will messages are no longer being sent. We tried v3.0.1 and it has the same issue. I see https://github.com/emqx/emqx/pull/2156 changes the behavior of will messages to not fire in the case of a clean disconnect, but did it remove will messages all together?

As far as I can tell, our client (https://github.com/eclipse/paho.mqtt.javascript) should not be cleanly disconnecting when a browser tab is closed.

**Comment by HJianBo** [comment]
Nope, we just modify the will message implement to follow the [MQTT protocol - Will Flag](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc385349230)

If the Javascript client send the DISCONNECT packet after a browser tab closed, Broker should not deliver Will Message

**Comment by HJianBo** [comment]
Nope, we just modify the will message implement to follow the [MQTT protocol - Will Flag](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc385349230)

If the Javascript client send the DISCONNECT packet after a browser tab closed, Broker should not deliver Will Message

**Comment by kleisauke** [comment]
I've the same issue with EMQ X v3.0.1 and v3.1.1. Looks like the LWT message is being ignored by the broker when a connection is made via WebSockets. I don't see this problem with TCP connections.

I tested both [paho.mqtt.javascript](https://github.com/eclipse/paho.mqtt.javascript) and [MQTT.js](https://github.com/mqttjs/MQTT.js) clients. Things I tried:
- Closing the browser tab.
- Closing the browser.
- Disconnected my network adapter.

None of the above results in an LWT message. Reproduction gist: https://gist.github.com/kleisauke/632b0f687bc9762e911d901e09e0026a

(just change `<BROKER URL HERE>` with the broker URL and run `subscribe-will.html` on another PC)

**Comment by mariusstaicu** [comment]
I can confirm this behavior also.

**Comment by kleisauke** [comment]
Quick update:
I can confirm that it works properly with EMQ X v3.0.0. The changes from PR #2156 indeed introduced this bug (thanks to `git bisect`).

Applying this patch on master (44d901e) seems to solve this issue:
<details>
 <summary>Patch</summary>

```diff
diff --git a/src/emqx_session.erl b/src/emqx_session.erl
index 1111111..2222222 100644
--- a/src/emqx_session.erl
+++ b/src/emqx_session.erl
@@ -617,21 +617,11 @@ handle_info({'EXIT', ConnPid, Reason}, #state{conn_pid = ConnPid})
     exit(Reason);
 
 handle_info({'EXIT', ConnPid, Reason}, State = #state{will_msg = WillMsg, expiry_interval = 0, conn_pid = ConnPid}) ->
-    case Reason of
-        normal ->
-            ignore;
-        _ ->
-            send_willmsg(WillMsg)
-    end,
+    send_willmsg(WillMsg),
     {stop, Reason, State#state{will_msg = undefined, conn_pid = undefined}};
 
 handle_info({'EXIT', ConnPid, Reason}, State = #state{conn_pid = ConnPid}) ->
-    State1 = case Reason of
-                 normal ->
-                     State#state{will_msg = undefined};
-                 _ ->
-                     ensure_will_delay_timer(State)
-             end,
+    State1 = ensure_will_delay_timer(State),
     {noreply, ensure_expire_timer(State1#state{conn_pid = undefined})};
 
 handle_info({'EXIT', OldPid, _Reason}, State = #state{old_conn_pid = OldPid}) ->
```
</details>

This partly reverts the changes of PR #2156, so it will still send the LWT message when the client sends the DISCONNECT packet with reason code 0. It needs proper handling of WebSocket disconnections, which this patch doesn't provide.

**Comment by hgascoigne** [comment]
Still having this issue on v3.2-beta.3. I tried upgrading from v3.0.0 to fix an issue with the broker timing out while trying to forward messages (which upgrading appears to fix), but now I'm running into this issue again.

I'd normally be happy to submit a PR but I'm pretty unfamiliar with erlang and would probably end up causing more harm than good :).

**Comment by Gilbert-Wong** [comment]
Hi, everyone, this bug has been resolved in the emqx v3.2.0 release. Here is the [pull request](https://github.com/emqx/emqx/pull/2680). Welcome to try emqx v3.2.0 :)

**Comment by kleisauke** [comment]
I can confirm that it works properly with emqx v3.2.0, thanks!
